homie-influx/src/config.rs

```diff
@@ -1,105 +1,158 @@
+use eyre::Report;
 use influx_db_client::reqwest::Url;
 use influx_db_client::Client;
 use rumqttc::MqttOptions;
 use rustls::ClientConfig;
+use serde::{de, Deserialize, Deserializer};
 use stable_eyre::eyre;
 use stable_eyre::eyre::WrapErr;
-use std::fs::File;
-use std::io::{BufRead, BufReader};
+use std::fmt::Display;
+use std::fs::read_to_string;
+use std::str::FromStr;
 use std::sync::Arc;
 
 const DEFAULT_MQTT_CLIENT_PREFIX: &str = "homie-influx";
 const DEFAULT_MQTT_HOST: &str = "test.mosquitto.org";
 const DEFAULT_MQTT_PORT: u16 = 1883;
 const DEFAULT_INFLUXDB_URL: &str = "http://localhost:8086";
-const DEFAULT_MAPPINGS_FILENAME: &str = "mappings.conf";
+const CONFIG_FILENAME: &str = "homie_influx.toml";
+const DEFAULT_MAPPINGS_FILENAME: &str = "mappings.toml";
+
+#[derive(Clone, Debug, Default, Deserialize)]
+#[serde(default)]
+pub struct Config {
+    pub mqtt: MqttConfig,
+    pub homie: HomieConfig,
+    pub influxdb: InfluxDBConfig,
+}
+
+impl Config {
+    pub fn from_file() -> Result<Config, Report> {
+        let config_file = read_to_string(CONFIG_FILENAME)
+            .wrap_err_with(|| format!("Reading {}", CONFIG_FILENAME))?;
+        Ok(toml::from_str(&config_file)?)
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(default)]
+pub struct MqttConfig {
+    pub host: String,
+    pub port: u16,
+    pub use_tls: bool,
+    pub username: Option<String>,
+    pub password: Option<String>,
+    pub client_prefix: String,
+}
+
+impl Default for MqttConfig {
+    fn default() -> MqttConfig {
+        MqttConfig {
+            host: DEFAULT_MQTT_HOST.to_owned(),
+            port: DEFAULT_MQTT_PORT,
+            use_tls: false,
+            username: None,
+            password: None,
+            client_prefix: DEFAULT_MQTT_CLIENT_PREFIX.to_owned(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(default)]
+pub struct HomieConfig {
+    pub mappings_filename: String,
+}
+
+impl Default for HomieConfig {
+    fn default() -> HomieConfig {
+        HomieConfig {
+            mappings_filename: DEFAULT_MAPPINGS_FILENAME.to_owned(),
+        }
+    }
+}
+
+#[derive(Clone, Debug, Deserialize)]
+#[serde(default)]
+pub struct InfluxDBConfig {
+    #[serde(deserialize_with = "de_from_str")]
+    pub url: Url,
+    pub username: Option<String>,
+    pub password: Option<String>,
+}
+
+/// Deserialize a FromStr by deserializing it as a string then parsing it.
+fn de_from_str<'de, D: Deserializer<'de>, T: FromStr>(d: D) -> Result<T, D::Error>
```

Comment:
This is used exactly once. Why is it so mental?

Reply:
I couldn't see any utility that already existed for this, so I wrote my own.

Reply:
(I wonder if the Url type is secretly already deserialisable if you apply the right feature flags or something. Is it the same type as the one mentioned in here https://github.com/servo/rust-url/issues/420#issuecomment-512239238 or it it reqwest specific?)

Reply:
Turns out it does! That's nicer.